insonmnia/oracle/oracle.go

```diff
@@ -0,0 +1,179 @@
+		zap.String("account", crypto.PubkeyToAddress(key.PublicKey).String()),
+		zap.String("price update period:", cfg.Oracle.PriceUpdatePeriod.String()),
+		zap.String("contract update period", cfg.Oracle.ContractUpdatePeriod.String()),
+		zap.Float64("deviation percent", cfg.Oracle.Percent))
+
+	bch, err := blockchain.NewAPI(blockchain.WithConfig(cfg.Blockchain))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Oracle{
+		logger:       logger,
+		key:          key,
+		cfg:          cfg,
+		bch:          bch,
+		currentPrice: big.NewInt(0),
+		actualPrice:  big.NewInt(0),
+	}, nil
+}
+
+func (o *Oracle) Serve(ctx context.Context) error {
+	priceWatcher := NewPriceWatcher(o.cfg.Oracle.PriceUpdatePeriod)
+	dw := priceWatcher.Start(ctx)
+
+	t := util.NewImmediateTicker(o.cfg.Oracle.ContractUpdatePeriod)
+
+	events, err := o.bch.Events().GetEvents(ctx, blockchain.GetMultiSigFilter([]common.Address{blockchain.OracleUsdMultisigAddr()}, big.NewInt(0)))
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case p := <-dw:
+			if p.err != nil {
+				return p.err
+			}
+			o.actualPrice = p.price
+			o.logger.Debug("loaded new price", zap.String("price", p.price.String()))
+		case <-t.C:
+			currentPrice, err := o.bch.OracleUSD().GetCurrentPrice(ctx)
+			if err != nil {
+				return err
+			}
+			if o.currentPrice.Cmp(currentPrice) != 0 {
+				o.currentPrice = currentPrice
+				o.logger.Debug("current price changed", zap.String("price", o.currentPrice.String()))
+			}
+			// master mode
+			if o.cfg.Oracle.Mode {
+				go o.SetPrice(ctx)
+			}
+		case event, ok := <-events:
+			if !ok {
+				return fmt.Errorf("events chanel closed")
+			}
+			// slave mode
+			if !o.cfg.Oracle.Mode {
+				o.logger.Debug("new submission", zap.Any("event", event))
+				switch value := event.Data.(type) {
+				case *blockchain.SubmissionData:
+					if o.transactionValid(ctx, value.TransactionId) {
+						go o.confirmChanging(ctx, value.TransactionId)
+					}
+				}
+			}
+		}
+	}
+}
+
+func (o *Oracle) SetPrice(ctx context.Context) error {
+	if o.actualPrice == nil {
+		o.logger.Debug("set price dropped, actual price not downloaded")
+		return fmt.Errorf("price is nil")
+	}
+	o.logger.Info("submitting new price", zap.String("price", o.actualPrice.String()))
+	data, err := o.bch.OracleUSD().PackSetCurrentPriceTransactionData(o.actualPrice)
+	if err != nil {
+		return err
+	}
+	return o.bch.OracleMultiSig().SubmitTransaction(ctx, o.key, blockchain.OracleUsdAddr(), big.NewInt(0), data)
+}
+
+// checkPrice check that submitted price does not differ more than 1%
+// incapsulate ugly big math to separate function
+func (o *Oracle) checkPrice(price *big.Int) bool {
+	diff := big.NewFloat(0).SetInt(o.actualPrice)
+	diff.Sub(diff, big.NewFloat(0).SetInt(price))
+	diff.Abs(diff)
+
+	p := big.NewInt(0).Set(o.actualPrice)
+	p.Div(p, big.NewInt(100))
+	percent := big.NewFloat(o.cfg.Oracle.Percent)
+	percent.Mul(percent, big.NewFloat(0).SetInt(p))
+
+	if diff.Cmp(percent) == -1 {
```

Comment:
what if you want to increase the price by exactly one percent?

Reply:
its big.int percent defined in config